lib/variable/reduction.cpp

```diff
@@ -22,148 +23,144 @@ using namespace scipp::core;
 namespace scipp::variable {
 
 namespace {
-
-Variable make_accumulant(const Variable &var, const Dim dim,
-                         const FillValue &init) {
-  if (variableFactory().has_masks(var))
-    throw except::NotImplementedError(
-        "Reduction operations for binned data with "
-        "event masks not supported yet.");
+/// Make a variable `accum` to pass to `sum_into(accum, data)`, etc.
+/// The variable is based on `data` and has the given dims.
+/// If `data` is binned, `accum` is dense with the elem dtype of `data`.
+Variable make_reduction_accumulant(const Variable &data,
+                                   const Dimensions &target_dims,
+                                   const FillValue init) {
+  const auto type = variableFactory().elem_dtype(data);
+  const auto unit = variableFactory().elem_unit(data);
+  const auto has_variances = variableFactory().has_variances(data);
+  const auto scalar_prototype = empty(Dimensions{}, unit, type, has_variances);
```

Comment:
I am looking at the sibling code in `dataset` now, https://github.com/scipp/scipp/blob/187e816849836acf060930e309479627a43130af/lib/dataset/variable_reduction.cpp#L19-L30.

There `zero_like` is used, but I think this would go wrong if applied to binned data, i.e., the approach here is the correct one. Is there a missing test in the data array with masks test?

Reply:
Try
```python
import scipp as sc

da = sc.data.binned_x(100, 3)
da.masks['mask'] = sc.array(dims=['x'], values=[False, False, True])
da.sum()
```
This works without the mask, with mask I get (on `main`):

```python
---------------------------------------------------------------------------
BinnedDataError                           Traceback (most recent call last)
Input In [6], in <cell line: 9>()
      7 da = sc.data.binned_x(100, 3)
      8 da.masks['mask'] = sc.array(dims=['x'], values=[False, False, True])
----> 9 da.sum()

File ~/software/scipp/scipp/core/reduction.py:112, in sum(x, dim)
     85 """Sum of elements in the input.
     86 
     87 If the input data is in single precision (dtype='float32') this internally uses
   (...)
    109     Ignore NaN's when calculating the sum.
    110 """
    111 if dim is None:
--> 112     return _cpp.sum(x)
    113 else:
    114     return _cpp.sum(x, dim=dim)

BinnedDataError: Binary operations such as '+' with binned data are only supported with dtype=VariableView, got dtype=DataArrayView. See https://scipp.github.io/user-guide/binned-data/computation.html#Event-centric-arithmetic for equivalent operations for binned (event) data.
```

Reply:
Fixed